tests/0056-balanced_group_mt.c

```diff
@@ -0,0 +1,309 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+
+/* Typical include path would be <librdkafka/rdkafka.h>, but this program
+ * is built from within the librdkafka source tree and thus differs. */
+#include "rdkafka.h" /* for Kafka driver */
+
+/**
+ * KafkaConsumer balanced group with multithreading tests
+ *
+ * Runs a consumer subscribing to a topic with multiple partitions and farms
+ * consuming of each partition to a separate thread.
+ */
+
+#define MAX_THRD_CNT 4
+
+static int assign_cnt = 0;
+static int consumed_msg_cnt = 0;
+static int consumers_running = 0;
+static int exp_msg_cnt;
+
+static mtx_t lock;
+static thrd_t tids[MAX_THRD_CNT];
+
+typedef struct part_consume_info_s {
+    rd_kafka_queue_t * rkqu;
+    int partition;
+} part_consume_info_t;
+
+static int is_consuming () {
+        int result;
+        mtx_lock(&lock);
+        result = consumers_running;
+        mtx_unlock(&lock);
+        return result;
+}
+
+static int partition_consume (void *args) {
+        part_consume_info_t *info = (part_consume_info_t *)args;
+        rd_kafka_queue_t *rkqu = info->rkqu;
+        int partition = info->partition;
+        int64_t ts_start = test_clock();
+        int max_time = (test_session_timeout_ms + 3000) * 1000;
+        int running = 1;
+
+        free(args); /* Free the parameter struct dynamically allocated for us */
+
+        while (ts_start + max_time > test_clock() && running &&
+               is_consuming()) {
+                rd_kafka_message_t *rkmsg;
+
+                rkmsg = rd_kafka_consume_queue(rkqu, 500);
+
+                if (!rkmsg)
+                        continue;
+                else if (rkmsg->err == RD_KAFKA_RESP_ERR__PARTITION_EOF)
+                        running = 0;
+                else if (rkmsg->err) {
+                        mtx_lock(&lock);
+                        TEST_FAIL("Message error "
+                                  "(at offset %" PRId64 " after "
+                                  "%d/%d messages and %dms): %s",
+                                  rkmsg->offset, consumed_msg_cnt, exp_msg_cnt,
+                                  (int)(test_clock() - ts_start) / 1000,
+                                  rd_kafka_message_errstr(rkmsg));
+                        mtx_unlock(&lock);
+                } else {
+                        mtx_lock(&lock);
+                        ++consumed_msg_cnt;
+                        if (rkmsg->partition != partition)
+                                TEST_FAIL("Message consumed has partition %d "
+                                          "but we expected partition %d.",
+                                          rkmsg->partition, partition);
+                        mtx_unlock(&lock);
+                }
+
+                rd_kafka_message_destroy(rkmsg);
+
+                mtx_lock(&lock);
+                if (consumed_msg_cnt >= exp_msg_cnt) {
+                        mtx_unlock(&lock);
+                        static int once = 0;
```

Comment:
static isnt safe in MT context, use `static RD_TLS int once ..` or use your lock.